### PR TITLE
Remove check for subgraph with cycles

### PR DESCRIPTION
**Pull request by Kh4L**
## Description ##
Partitioning a graph with potential subgraphs that have a cycle with the input or output fails instead of ignoring these subgraphs. For instance, Faster-RCNN (available in the GluonCV model zoo) optimization failed. 

## Change ##

This PR simply removes the check that raise an exception if the subgraph candidate has a cycle, so LabelSubgraph can return false.

### CODE REVIEW

File: src/operator/subgraph/build_subgraph.cc [hunk 1]
```diff
@@ -306,14 +300,19 @@ void PreSelectSubgraphNodes(const nnvm::Graph& g, SubgraphSelectorV2Ptr subgraph
...
+  size_t n_excluded_nodes = 0;
   const size_t max_num_retry = simple_nodes.size() * simple_nodes.size();
   size_t count = 0;
   bool success = false;
   while (!success && count < max_num_retry) {
     success = LabelSubgraph(g, subgraph_selector, label, snid, simple_nodes, subgraph_nodes,
                             &excluded_nodes);
     if (!success) {
-      CHECK(!excluded_nodes.empty());
+      if (excluded_nodes.size() == n_excluded_nodes) {
```
Comment (samskalicky) [inline src/operator/subgraph/build_subgraph.cc:311]: Do nodes have to be excluded in order to have a possible subgraph?

File: src/operator/subgraph/build_subgraph.cc [hunk 2]
```diff
@@ -306,14 +300,19 @@ void PreSelectSubgraphNodes(const nnvm::Graph& g, SubgraphSelectorV2Ptr subgraph
...
+        // no possible subgraph for the current node snid
+        break;
+      }
+      n_excluded_nodes = excluded_nodes.size();
```
Comment (samskalicky) [inline src/operator/subgraph/build_subgraph.cc:315]: can we add a comment here, "failed to label subgraph due to loop, excluding nodes to remove loop and retrying"

File: src/operator/subgraph/build_subgraph.cc [hunk 3]
```diff
@@ -306,14 +300,19 @@ void PreSelectSubgraphNodes(const nnvm::Graph& g, SubgraphSelectorV2Ptr subgraph
...
   const size_t max_num_retry = simple_nodes.size() * simple_nodes.size();
   size_t count = 0;
   bool success = false;
   while (!success && count < max_num_retry) {
     success = LabelSubgraph(g, subgraph_selector, label, snid, simple_nodes, subgraph_nodes,
                             &excluded_nodes);
     if (!success) {
-      CHECK(!excluded_nodes.empty());
+      if (excluded_nodes.size() == n_excluded_nodes) {
+        // no possible subgraph for the current node snid
```
Comment (samskalicky) [inline src/operator/subgraph/build_subgraph.cc:312]: can we extend this comment to something like "failed to label subgraph to loop, excluding nodes does not remove loop"

**Comment by samskalicky** [comment]
Thanks @Kh4L! this is a really great addition to the partitioning logic to be able to skip subgraphs rather than error out when theres a loop. It took me a while to come up to speed with the complexity of the logic here, can you add some more comments so that next time it will be easier?

**Comment by Kh4L** [comment]
@mxnet-bot run ci [unix-cpu]

**Comment by Kh4L** [comment]
@mxnet-bot run ci [clang]

**Comment by Kh4L** [comment]
@mxnet-bot run ci [centos-cpu]

**Pull request merged**